public/js/src/module/store.js

```diff
@@ -668,32 +681,36 @@ function _firstItemOnly( results ) {
     }
 }
 
-function _transformExternalDataToXmlStr( survey ) {
-    if ( survey && survey.externalData ) {
-        survey.externalData = survey.externalData.map( instance => {
-            if ( instance.xml instanceof XMLDocument ) {
-                instance.xml = new XMLSerializer().serializeToString( instance.xml.documentElement, 'text/xml' );
-            }
+function _serializeSurvey( survey ) {
+    if ( survey ) {
+        if ( survey.externalData ) {
+            survey.externalData = survey.externalData.map( instance => {
+                if ( instance.xml instanceof XMLDocument ) {
+                    instance.xml = new XMLSerializer().serializeToString( instance.xml.documentElement, 'text/xml' );
+                }
 
-            return instance;
-        } );
-    }
+                return instance;
+            } );
+        }
 
-    return survey;
+        return encryptor.serializeEncryptedSurvey( survey );
+    }
 }
 
-function _transformExternalDataToXmlDoc( survey ) {
-    if ( survey && survey.externalData ) {
-        survey.externalData = survey.externalData.map( instance => {
-            if ( typeof instance.xml === 'string' ) {
-                instance.xml = parser.parseFromString( instance.xml, 'text/xml' );
-            }
+function _deserializeSurvey( survey ) {
```

Comment:
IndexedDB requires any stored value to be serialized in full, and while many data types are serializable some unfortunately aren't. I'm having trouble finding the original documentation I referenced but [this MDN doc](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) and [the W3C spec](https://html.spec.whatwg.org/multipage/structured-data.html#safe-passing-of-structured-data) go into some detail here.

`Symbol` (the mechanism used in `encryptor` to mark a Survey as having encryption enabled) is notably _not_ serializable. So this additional step to serialization transforms `Symbol('isEncryptionEnabled')` to a plain string property `isEncryptionEnabled` for storage, then back to the symbol when retrieved in the runtime.

My feeling was that it was safer to have a single point for de/serialization of Survey than add another field-specific set of private functions which would _always_ be paired with the existing one.

Reply:
In the fullness of time, I would prefer to separate concerns like this entirely with clearly delineated runtime (and type safe) validators/transformers not unlike [io-ts](https://github.com/gcanti/io-ts) or [Runtypes](https://github.com/pelotom/runtypes).

Reply:
I see, that all makes sense. I'm still not a huge fan of the name because it feels like it's not clear what is being deserialized or why but don't have an alternative to propose at this time. Maybe a quick comment here would help provide context?

Reply:
I'm not quite convinced that the Symbol (for setting isEncryptionEnabled) adds meaningful security but I may very well be overlooking something. My thinking is that if somebody wants to prevent encryption from happening they'd already have access to the not-yet-encrypted instance. Wouldn't changing this setting be the same level of effort as simply accessing the model before the record is encrypted?